tradeoffs.Rmd

```diff
@@ -0,0 +1,172 @@
+#  (PART) Conclusion {-}
+
+# Tradeoffs
+
+::: {.alert .alert-success} 
+This somewhat opinionated discussion was added by [Max Held](https://www.maxheld.de).
+Your mileage may vary.
+:::
+
+Test doubles such as mocks or fakes can help you isolate your tests from the state of external HTTP API.
+This can be a good thing, because it allows you to test your code against a rare responses, or when the API is not available.
+
+Unfortunately, mocking and faking add their own complexity to a project.
+Respectively, you have to carry expected responses or even (rudimentary) API behavior in your code base.
+You now have two relationships to keep in mind: your code must pass your mocked/faked tests, and your mocks/fakes must correspond to the live API.
+
+This tradeoff cannot be resolved, but it is eased considerably by decoupled (or modular) code.
+Much as with other side effects, it can help to concentrate your API calls in as few functions as possible and test these with mocks or fakes, where necessary.
+All other functions can then be tested without dependence on the API.
+Conversely, if you find yourself requiring these test doubles in a lot of places, you may want to [consider a refactoring](https://medium.com/javascript-scene/mocking-is-a-code-smell-944a70c90a6a).
+
+## Unit vs Integration Testing
+
+Test doubles also benefit from a clean separation of concerns.
+
+On the one hand, [*unit tests*](https://en.wikipedia.org/wiki/Unit_testing) should only test the behavior of your code, not the behavior of the API.
+Otherwise a failed test could be ambiguous: Was it your code, or a errant API behavior?
+Unit tests will typically cover simple, but user facing behavior, such as whether an API response is parsed correctly, or whether an HTTP error code is handled appropriately.
+Mocks, especially, can help you write isolated unit tests.
+
+On the other hand, [*integration tests*](https://en.wikipedia.org/wiki/Integration_testing) should test how your code interacts with other software or services.
+["Narrow" integration tests](https://martinfowler.com/bliki/IntegrationTest.html) are run against test doubles, typically fakes.
+This may be most helpful to test more complex, perhaps stateful, interactions between your code and the API, which relies on internal logic of the API rather than a one-off response.
+For example, you may want to test whether if you submit a string to a translation service, you get back a result in the appropriate language.
```

Comment:
Citing these notions is interesting for sure, but it would help to link to real R packages using them otherwise the chapter might be too dry, what do you think?

Reply:
I take up these notions in the below example, and I think it's just helpful vocabulary.
I'd hope that you can find these distinctions in some R packages, though I wasn't able to find a clean-cut example. 
Also seems to me that explaining the context of another API wrapper might bloat the chapter more.

Reply:
But even in a few words, what would be a narrow integration test for an R package?

Reply:
ah, I see, yes, that sort of illustration really is needed.